### PR TITLE
nsys-jax: time-alignment and lock-step analysis

### PR DESCRIPTION
**Pull request by olupton**
Also:
- use "device" instead of "rank" everywhere
- move warmup removal heuristics to standalone function

### CODE REVIEW

File: .github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py [hunk 1]
```diff
@@ -3,13 +3,111 @@
...
+    init = {"compile": frames["compile"]}
+    steady = {}
+    for k in ["communication", "thunk", "module"]:
+        if k not in frames:
+            continue
+        df = frames[k]
+        compile_mask = df.index.get_level_values("ProgramId").isin(compilation_ids_seen)
+        prog_exec_values = df.index.get_level_values("ProgramExecution")
+        init_mask = compile_mask & (prog_exec_values == 0)
+        steady_mask = ~compile_mask | (compile_mask & (prog_exec_values > 1))
```
Comment (gspschmid) [inline .github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py:104]: Nit: This should be equivalent to `~compile_mask | (prog_exec_values > 1)`, no?

File: .github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py [hunk 2]
```diff
@@ -3,13 +3,111 @@
...
+        assert (prog_exec_values[~init_mask & ~steady_mask] == 1).all()
```
Comment (gspschmid) [inline .github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py:105]: Maybe also add an assertion that we got any entries in steady_mask? (To notify the user of situations in which they didn't profile long enough to get any useful data.)

File: .github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py [hunk 3]
```diff
@@ -3,13 +3,111 @@
...
+    # Apply these corrections to the device-side timestamps
+    for k in ["communication", "module", "thunk"]:
+        df = frames[k]
+        df["ProjStartNs"] -= median_device_skews
+        frames[k] = df
+    return frames, {
+        "collective_end_time_skews_ns": end_time_skews,
+        "device_corrections": median_device_skews,
+        "collective_size": max_collective_size,
+    }
```
Comment (gspschmid) [inline .github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py:60]: `align_profiler_data_timestamps` is really neat and concise!

**Review by gspschmid — APPROVED** [review]
A few minor comments

File: .github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py [hunk 4]
```diff
@@ -3,13 +3,111 @@
...
+from typing import Any
 
 from .protobuf import xla_module_metadata
 from .utils import make_child_mask
 
 pd.options.mode.copy_on_write = True
 
 
+def align_profiler_data_timestamps(
+    frames: dict[str, pd.DataFrame],
```
[the remark this answers is not in the document]
Comment (olupton) [reply .github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py]: Good suggestion, done :+1:

File: .github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py (hunk 1 above)
Comment (olupton) [reply .github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py]: Well spotted, thanks.

File: .github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py (hunk 2 above)
Comment (olupton) [reply .github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py:104]: Done

**Review by gspschmid — APPROVED** [review]

**Pull request merged**